Use C bindings in MPI example

effeffe opened the pull request:
This pull request simply rewrites the MPI interface to use C bindings instead of C++ bindings. This is because C++ bindings have been deprecated since OpenMPI version 3 and require explicit flags to compile; moreover these have been completely removed from version 5 onwards.

This simple rewrite allows for MPI usage on any modern system. It has been tested on Geant4 11.2.1 and Geant4 11.3.0 using both OpenMPI 4.1.4 and 5.0.2. Tests have been done on a laptop and on a SLURM cluster.

The parser of the manager has been slightly rewritten as well to be able to parse command line arguments into the program. This is especially useful to set any number of threads per node.
The default behaviour would be to create one MPI node ("virtual") per thread to allow the use of any number of threads in each node.

drbenmorgan commented:
Thanks for this submission! We're currently integrating a preexisting, independent patch that does the migration, so will try your fixes on top of this once merged upstream.

drbenmorgan commented:
Did your tests cover build and run of the `exMPI0N` examples, and if so, what invocation of `mpirun` did you use?

ihrivnac commented:
> Did your tests cover build and run of the `exMPI0N` examples, and if so, what invocation of `mpirun` did you use?

I followed the examples documentation:
https://geant4-userdoc.web.cern.ch/Doxygen/examples_doc/html/exMPI03.html

And  run with the `mpiexec` command: 
```bash
mpiexec -n # ./exMPI03 [run.mac]
```

effeffe commented:
> Did your tests cover build and run of the `exMPI0N` examples, and if so, what invocation of `mpirun` did you use?

I have tested exMPI01 and exMPI04 and they both worked fine.
On my laptop I used
``` $ mpirun -n $CORES exMPI0# ```
On slurm I didn't even specify the number of nodes. It automatically picked up the number of theeads as a single core each. While this might mot be optimal, it allows for flexibility as the scheduler can give you any amount of cores in each node.

effeffe commented on examples/extended/parallel/MPI/source/CMakeLists.txt:
I would probably edit this to condition Geant4 10 or 11 and later

ihrivnac commented:
I tried `mpirun` instead of `mpiexec` and the result is same.  The program hangs after printing from vizualization:

```
2:You have successfully registered the following user vis actions.
2:Run Duration User Vis Actions: none
2:End of Event User Vis Actions: none
2:End of Run User Vis Actions: none
2:
2:Some /vis commands (optionally) take a string to specify colour.
"/vis/list" to see available colours.
2:
```

ihrivnac commented:
I am sorry, I have mismatched with the parallel discussions ongoing on gitlab which @bmorgan mentioned above. I was testing the MR in the gitlab and not yours !

effeffe commented:
Sorry my bad. I tested exMPI01 and exMPI02. 03 and 04 need a partial rewrite in the runMerger, as C++ bindings are present there too

effeffe commented:
After this last commit, the macros get parsed even outside the macro folder. I have attached the terminal outputs

[run_exMPI01.txt](https://github.com/user-attachments/files/18867438/run_exMPI01.txt)
[run_exMPI02.txt](https://github.com/user-attachments/files/18867439/run_exMPI02.txt)
[run_exMPI03.txt](https://github.com/user-attachments/files/18867442/run_exMPI03.txt)
[run_exMPI04.txt](https://github.com/user-attachments/files/18867444/run_exMPI04.txt)

drbenmorgan commented:
Thanks for the updates, these fix the hanging and the non-MPI and MPI runs at least with 2 instances run. There are a few compiler warnings in G4mpi related to shadowing and unused variables with GCC 11 and 13. Could you fix these please?

drbenmorgan commented on examples/extended/parallel/MPI/source/include/G4MPIscorerMerger.hh:
```diff
@@ -103,8 +103,9 @@ class G4MPIscorerMerger
     unsigned int commSize;
     unsigned int destinationRank;
-    MPI::Intracomm comm;
+    MPI_Comm comm;  // Changed from MPI::Intracomm to MPI_Comm
```
```suggestion
    MPI_Comm comm;
```

No need to add a comment here.

drbenmorgan commented on examples/extended/parallel/MPI/source/src/G4MPIhistoMerger.cc:
```diff
@@ -52,8 +52,9 @@ void G4MPIhistoMerger::Merge()
-  const MPI::Intracomm* parentComm = G4MPImanager::GetManager()->GetComm();
-  MPI::Intracomm comm = parentComm->Dup();
+  const MPI_Comm* parentComm = G4MPImanager::GetManager()->GetComm();
+  MPI_Comm comm;// = parentComm->Dup();
```
```suggestion
  MPI_Comm comm;
```
No need for comment.

drbenmorgan commented on examples/extended/parallel/MPI/source/src/G4MPImanager.cc:
```diff
@@ -84,7 +85,10 @@ G4MPImanager::G4MPImanager(int nof_extra_workers)
-  MPI::Init_thread(MPI::THREAD_SERIALIZED);
+  MPI_Init_thread(nullptr, nullptr, MPI_THREAD_SERIALIZED, &provided_);
+  if (provided_ < MPI_THREAD_SERIALIZED) {
+    std::cerr << "Warning: MPI does not provide the requested threading support!" << std::endl;
```
```suggestion
    G4cout << "Warning: MPI does not provide the requested threading support!" << G4endl;
```

What's the behaviour if this condition is true? Will the application still run, or is it a hard error condition?

effeffe replied:
I'm not too sure, I will need to test it out

drbenmorgan replied:
If you could, that'd be great to clarify this.

effeffe replied:
I have been trying to trigger this condition on my machine, but without actual success. I have only been able to trigger this replacing provided_ with MPI_THREAD_SINGLE.

In theory, the application should not run if there are problems with the MPI threading level, so it should be left as such. We definitely don't want anything below SERIALISED. As least, this is the strong suggestion that the documentation makes: https://docs.open-mpi.org/en/v5.0.x/man-openmpi/man3/MPI_Init_thread.3.html#mpi-init-thread

drbenmorgan replied:
Thanks for investigating! Given that running with a level below `SERIALIZED` isn't good, this should be reported as a fatal `G4Exception` to terminate the application at this point. Something like:

```c++
int provided_;
MPI_Init_thread(nullptr, nullptr, MPI_THREAD_SERIALIZED, &provided_);
if(provided_ < MPI_THREAD_SERIALIZED)
{
  G4Exception("G4MPImanager::G4MPImanager()",
              "G4MPImanager001",
               FatalException,
	       "MPI Initialization failed to setup with MPI_THREAD_SERIALIZED or better");
}
```

effeffe replied:
Good idea, fixed!

drbenmorgan commented on examples/extended/parallel/MPI/source/src/G4MPImanager.cc:
```diff
@@ -173,14 +181,14 @@ void G4MPImanager::Initialize()
 
   // get rank information
-  world_size_ = MPI::COMM_WORLD.Get_size();
+  MPI_Comm_size(MPI_COMM_WORLD, &world_size_);//world_size_ = MPI::COMM_WORLD.Get_size();
```
```suggestion
  MPI_Comm_size(MPI_COMM_WORLD, &world_size_);
```

Remove trailing comment

drbenmorgan commented on examples/extended/parallel/MPI/source/src/G4MPImanager.cc:
```diff
@@ -173,14 +181,14 @@ void G4MPImanager::Initialize()
   }
   size_ = world_size_ - nof_extra_workers_;
-  rank_ = MPI::COMM_WORLD.Get_rank();
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank_); //rank_ = MPI::COMM_WORLD.Get_rank();
```
```suggestion
  MPI_Comm_rank(MPI_COMM_WORLD, &rank_);
```

Remove trailing comment.

drbenmorgan commented on examples/extended/parallel/MPI/source/src/G4MPIscorerMerger.cc:
```diff
@@ -168,14 +168,15 @@ void G4MPIscorerMerger::Merge()
   handler_t sender = std::bind(&G4MPIscorerMerger::Send, this, _1);
   handler_t receiver = std::bind(&G4MPIscorerMerger::Receive, this, _1);
-  std::function<void(void)> barrier = std::bind(&MPI::Intracomm::Barrier, &comm);
+  //std::function<void(void)> barrier = std::bind(MPI_Barrier, comm);
```
```suggestion
```

Remove extraneous comment.

ihrivnac commented on examples/extended/parallel/MPI/source/src/G4MPImanager.cc:
```diff
@@ -84,7 +85,10 @@ G4MPImanager::G4MPImanager(int nof_extra_workers)
 {
   // MPI::Init();
-  MPI::Init_thread(MPI::THREAD_SERIALIZED);
+  MPI_Init_thread(nullptr, nullptr, MPI_THREAD_SERIALIZED, &provided_);
```
There is no need to introduce a data member `provided_` as this variable is used only in the following test and nowhere else in the class.

effeffe replied:
Should we reinstroduce the `if (provided_ < MPI_THREAD_SERIALIZED)` test instead?

drbenmorgan replied:
I think it's now o.k. using just a local variable in this member function.

ihrivnac commented on examples/extended/parallel/MPI/source/src/G4MPImanager.cc:
```diff
@@ -102,8 +106,12 @@ G4MPImanager::G4MPImanager(int argc, char** argv, int nof_extra_workers)
-  // MPI::Init(argc, argv);
-  MPI::Init_thread(argc, argv, MPI::THREAD_SERIALIZED);
+  int provided;
+  MPI_Init_thread(nullptr, nullptr, MPI_THREAD_SERIALIZED, &provided_);
```
Same as above

ihrivnac commented on examples/extended/parallel/MPI/source/include/G4MPImanager.hh:
```diff
@@ -123,11 +124,14 @@ class G4MPImanager
     const MPI_Comm* GetCollectingComm() const { return &collecting_comm_; }
     const MPI_Comm* GetAllComm() const { return &all_comm_; }
 
+    std::vector<G4String> ReturnArguments() { return _options; }
```
This function seems to be redundant, as it is not used anywhere in `source` or `examples`

effeffe replied:
This actually is an improvement on the framework. It allows to return the command line arguments, so that the main program can use it.

drbenmorgan replied:
It's really a design error that `G4MPImanager` ever parsed/consumed command line arguments in the first place (other than forwarding them to the MPI API). A longer term solution is needed, which is outside the scope of this PR, but here and for now we shouldn't be adding functionality that won't be supported long term.

effeffe replied:
I understand. But should we keep it as proposed so at least it's working where needed, or should we remove the function and effectively break functionality until someone fixes this?

I understand the principle of not messing up userspace, but this effectively fixes something that was broken before.

For instance, the G4MTRunManager doesn't always get the correct number of threads, and as such it is important to be able to modify this from command line where necessary. This function effectively allows adding -t $THREADS to the command line in order to pass it to `runManager->SetNumberOfThreads(NumberOfThreads);` .

ihrivnac replied:
Thanks @effeffe for yor explications; it's fine with me to keep the function in this PR.

drbenmorgan commented on examples/extended/parallel/MPI/source/include/G4MPIscorerMerger.hh:
```diff
@@ -103,8 +103,9 @@ class G4MPIscorerMerger
-    MPI::Intracomm comm;
+    MPI_Comm comm;
     G4int verbose;
+    void G4mpi_barrier(MPI_Comm* comm) {MPI_Barrier(*comm);};
```
This function also appears redundant, at least not used anywhere else?

ihrivnac commented on examples/extended/parallel/MPI/source/include/G4MPIscorerMerger.hh:
```diff
@@ -2,24 +2,24 @@
-// * work  make  any representation or  warranty, express or implied, *
-// * regarding  this  software system or assume any liability for its *
-// * use.  Please see the license in the file  LICENSE  and URL above *
+// * institutes, nor the agencies providing financial support for this *
+// * work make any representation or warranty, express or implied, *
+// * regarding this software system or assume any liability for its *
+// * use. Please see the license in the file LICENSE and URL above *
 // * for the full disclaimer and the limitation of liability.         *
 // *                                                                  *
-// * This  code  implementation is the result of  the  scientific and *
+// * This code implementation is the result of the scientific and *
 // * technical work of the GEANT4 collaboration.                      *
-// * By using,  copying,  modifying or  distributing the software (or *
-// * any work based  on the software)  you  agree  to acknowledge its *
-// * use  in  resulting  scientific  publications,  and indicate your *
+// * By using, copying, modifying or distributing the software (or *
+// * any work based on the software) you agree to acknowledge its *
+// * use in resulting scientific publications, and indicate your *
 // * acceptance of all terms of the Geant4 Software license.          *
```
The License text should not be modified

drbenmorgan commented on examples/extended/parallel/MPI/source/include/G4MPImanager.hh:
```diff
@@ -153,11 +157,12 @@ class G4MPImanager
     G4bool is_slave_;
     G4bool is_extra_worker_;
     G4int rank_;
+    G4int provided_;
```
```suggestion
```

I think this can now be removed as the latest commits use function local variables instead.

drbenmorgan commented:
Great! I'll get started on getting this tested upstream and reconciled with the other private patch.

effeffe commented:
> warning

I was trying to have a look at them, bu tI was not getting any. I have added
```
if (MSVC)
    # warning level 4
    add_compile_options(/W4)
else()
    # additional warnings
    add_compile_options(-Wall -Wextra -Wpedantic)
endif()
```
I have fixed a warning related to an unused variable in parser().
I still get the following warnings:
```
filippo@iamyoubutArch> make -j1
[  5%] Building CXX object CMakeFiles/G4mpi.dir/analysis/src/G4RootMpiAnalysisManager.cc.o
[ 10%] Building CXX object CMakeFiles/G4mpi.dir/analysis/src/G4RootMpiNtupleFileManager.cc.o
[ 15%] Building CXX object CMakeFiles/G4mpi.dir/analysis/src/G4RootMpiNtupleManager.cc.o
[ 20%] Building CXX object CMakeFiles/G4mpi.dir/analysis/src/G4RootMpiPNtupleManager.cc.o
[ 25%] Building CXX object CMakeFiles/G4mpi.dir/src/G4MPIbatch.cc.o
[ 30%] Building CXX object CMakeFiles/G4mpi.dir/src/G4MPIextraWorker.cc.o
[ 35%] Building CXX object CMakeFiles/G4mpi.dir/src/G4MPIhistoMerger.cc.o
[ 40%] Building CXX object CMakeFiles/G4mpi.dir/src/G4MPImanager.cc.o
/home/filippo/Documents/Geant4/MPI/source/src/G4MPImanager.cc: In member function ‘void G4MPImanager::ExecuteBeamOnThread(const G4String&)’:
/home/filippo/Documents/Geant4/MPI/source/src/G4MPImanager.cc:535:47: warning: cast between incompatible function types from ‘void (*)(const G4String*)’ to ‘Func_t’ {aka ‘void* (*)(void*)’} [-Wcast-function-type]
  535 |     G4int rc = pthread_create(&thread_id_, 0, (Func_t)thread_ExecuteThreadCommand, (void*)&cmdstr);
      |                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[ 45%] Building CXX object CMakeFiles/G4mpi.dir/src/G4MPImessenger.cc.o
[ 50%] Building CXX object CMakeFiles/G4mpi.dir/src/G4MPIntupleMerger.cc.o
[ 55%] Building CXX object CMakeFiles/G4mpi.dir/src/G4MPIrandomSeedGenerator.cc.o
[ 60%] Building CXX object CMakeFiles/G4mpi.dir/src/G4MPIscorerMerger.cc.o
[ 65%] Building CXX object CMakeFiles/G4mpi.dir/src/G4MPIsession.cc.o
[ 70%] Building CXX object CMakeFiles/G4mpi.dir/src/G4MPIstatus.cc.o
[ 75%] Building CXX object CMakeFiles/G4mpi.dir/src/G4MPIutils.cc.o
In file included from /usr/include/c++/14.2.1/x86_64-pc-linux-gnu/bits/c++allocator.h:33,
                 from /usr/include/c++/14.2.1/bits/allocator.h:46,
                 from /usr/include/c++/14.2.1/bits/alloc_traits.h:39,
                 from /usr/include/c++/14.2.1/ext/alloc_traits.h:34,
                 from /usr/include/c++/14.2.1/bits/hashtable_policy.h:39,
                 from /usr/include/c++/14.2.1/bits/hashtable.h:35,
                 from /usr/include/c++/14.2.1/bits/unordered_map.h:33,
                 from /usr/include/c++/14.2.1/unordered_map:41,
                 from /usr/include/c++/14.2.1/functional:63,
                 from /home/filippo/Documents/Geant4/MPI/source/include/G4MPIutils.hh:29,
                 from /home/filippo/Documents/Geant4/MPI/source/src/G4MPIutils.cc:26:
In member function ‘void std::__new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = unsigned int; _Args = {const unsigned int&}; _Tp = unsigned int]’,
    inlined from ‘static void std::allocator_traits<std::allocator<_Tp1> >::construct(allocator_type&, _Up*, _Args&& ...) [with _Up = unsigned int; _Args = {const unsigned int&}; _Tp = unsigned int]’ at /usr/include/c++/14.2.1/bits/alloc_traits.h:575:17,
    inlined from ‘void std::vector<_Tp, _Alloc>::push_back(const value_type&) [with _Tp = unsigned int; _Alloc = std::allocator<unsigned int>]’ at /usr/include/c++/14.2.1/bits/stl_vector.h:1288:30,
    inlined from ‘G4mpi::commMap_t G4mpi::buildCommunicationMap(std::vector<unsigned int>&)’ at /home/filippo/Documents/Geant4/MPI/source/src/G4MPIutils.cc:78:26:
/usr/include/c++/14.2.1/bits/new_allocator.h:191:11: warning: ‘<unknown>[4611686018427387902]’ may be used uninitialized [-Wmaybe-uninitialized]
  191 |         { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[ 80%] Building CXX object CMakeFiles/G4mpi.dir/src/G4UImpish.cc.o
[ 85%] Building CXX object CMakeFiles/G4mpi.dir/src/G4VMPIseedGenerator.cc.o
[ 90%] Building CXX object CMakeFiles/G4mpi.dir/src/G4VMPIsession.cc.o
[ 95%] Building CXX object CMakeFiles/G4mpi.dir/src/G4VUserMPIrunMerger.cc.o
[100%] Linking CXX shared library libG4mpi.so
[100%] Built target G4mpi
make -j16 -j1  30.59s user 1.69s system 99% cpu 32.529 total
```

I'm not too sure about the second warning. Any idea?

About the first one, we could modify `thread_ExecuteThreadCommand` to be
```
void* thread_ExecuteThreadCommand(void* arg)
{
  const G4String* command = static_cast<const G4String*>(arg);
  G4MPImanager::GetManager()->ExecuteThreadCommand(*command);
  return nullptr;
}
```

and remove the casting on the line.

drbenmorgan commented:
@effeffe sorry for the delay in replying - I'm a bit busy with other tasks this week, so bear with me on this!

effeffe commented on examples/extended/parallel/MPI/source/src/G4MPImanager.cc:
```diff
@@ -257,50 +272,55 @@ void G4MPImanager::Initialize()
+    G4String arg = argv[i];
+
+    G4String sub = arg.substr(0,5);
+    G4String endsub = arg.substr(arg.length()-4);
```
These do not have a failsafe in case the arguments are short. To be fixed -> WIP

effeffe replied:
Fixed

drbenmorgan commented:
Apologies again for the delay in looking at this! We've merged it in with the pre-existing patch (there are only whitespace/name differences, but we want to ensure equal credit!), and tested upstream which passed in the detailed tests fine. Now merged in the development branch, so will be in the 11.4 beta in June, and the full release in December. Many thanks for your contribution!

effeffe commented:
> Apologies again for the delay in looking at this! We've merged it in with the pre-existing patch (there are only whitespace/name differences, but we want to ensure equal credit!), and tested upstream which passed in the detailed tests fine. Now merged in the development branch, so will be in the 11.4 beta in June, and the full release in December. Many thanks for your contribution!

Cool, thank you all!
